### PR TITLE
Add FPS compatible option in second mode of the animation editor snapping

### LINKED ISSUE #98609: Animation editor key snapping is rounded to some predetermined value instead of using exact snap value
### Tested versions

- Reproducible in : 4.4.dev3, 4.4.dev2
- Not reproducible in : 4.2.2.stable

### System information

Windows 11 - Godot v4.4.dev3

### Issue description

When you set time snap value in animation editor and you try to create or move key, the result key time value is rounded to some predetermined value instead of using exact number to snap.
Some examples:
- If snap value is 0.15, key is actually snapped to value 0.1429
- If snap value is 0.225 or 0.275, key is actually snapped to value 0.25
- If snap value is 0.3 or 0.4, key is actually snapped to value 0.3333333...
- If snap value is from 0.45 to 0.75, key is actually snapped to value 0.5
- If snap value is bigger than 0.75, key is actually snapped to value 1
The same rules apply for any integer value like 2.225 is snapped to 2.25, or 4.3 is snapped to 4.33333...

Note that timeline cursor doesn't have this issue and it snaps correctly to exact snap value, no matter how much decimal part is.

### Steps to reproduce

1.) Create AnimationPlayer, some Sprite or ColorRect, and create some simple moving animation for it
2.) Try to set snapping value to 0.25, 0.3, 0.425, 0.45 and 0.75 and try to create new key or move existing one and see that it snaps to incorrect value, while timeline cursor is moving correctly (when timeline cursor snap is enabled)

### Minimal reproduction project (MRP)

[MRP.zip](https://github.com/user-attachments/files/17541034/MRP.zip)

### PR DESCRIPTION
**Pull request by TokageItLab**
- Fixes https://github.com/godotengine/godot/issues/98609
- ~Supersedes https://github.com/godotengine/godot/pull/98826~ Now conflict is solved

Makes it optional to snap the FPS to a close integer FPS when in SecondMode.

image

Partially revert https://github.com/godotengine/godot/pull/97569 to disallow fractional FPS because KeyEditor's spin slider syncs with the FPS of the Snap setting, but KeyEditor's spin slider only allows integers, so if fractional FPS is used for snapping, KeyEditor's spin slider will break. Moreover, it is not consistent with this FPS compatible option.

Even if the KeyEditor spin slider allows fractional values, this will still cause problems due to lack of precision[^1], so the fundamental solution is to separate the FPS values for KeyEditor/Timeline from the step(snapping) values. In other words, it means that we need to add a FPS value as well as a step value to the Animation property, but I feel that it should be an editor-only property because it is too specific to the niche case of fractional FPS.

[^1]: Since the value of time in KeyEditor exceeds 10.0 usually, there is a lack of precision if we want to use a step of 0.0625 since it will be like 10.0625 or 100.0625

### CODE REVIEW

File: editor/animation_track_editor.cpp [hunk 1]
```diff
@@ -5160,9 +5180,8 @@ void AnimationTrackEditor::_update_step(double p_new_step) {
 	double step_value = p_new_step;
 	if (timeline->is_using_fps()) {
 		if (step_value != 0.0) {
-			// step_value must also be less than or equal to 1000 to ensure that no error accumulates due to interactions with retrieving values from inner range.
+			// Step_value must also be less than or equal to 1000 to ensure that no error accumulates due to interactions with retrieving values from inner range.
```
Comment (AhmedL-dev) [inline editor/animation_track_editor.cpp:5183]: `step_value` is the name of the variable, can keep it lowercase for variable name consistency.

**Review by SaracenOne — APPROVED** [review]
Reviewed at animation meeting 👍

File: editor/animation_track_editor.cpp [hunk 2]
```diff
@@ -6377,8 +6396,8 @@ void AnimationTrackEditor::goto_prev_step(bool p_from_mouse_event) {
 		return;
 	}
 	float anim_step = animation->get_step();
-	if (anim_step == 0.0) {
-		anim_step = 1.0;
+	if (anim_step == 0) {
+		anim_step = 1;
```
Comment (akien-mga) [inline editor/animation_track_editor.cpp:6400]: Not too important, but I think this change was unnecessary, we prefer to be explicit with float literals and use the `.0` decimal for whole values (same below).

**Comment by Repiteo** [comment]
Thanks!

File: editor/animation_track_editor.cpp [hunk 3]
```diff
@@ -7314,19 +7333,30 @@ void AnimationTrackEditor::_selection_changed() {
...
+	nearest_fps = 0;
+
 	if (step->get_value() <= 0) {
 		snap_unit = 0;
+		_update_nearest_fps_label();
 		return; // Avoid zero div.
 	}
 
 	if (timeline->is_using_fps()) {
+		_clear_selection(true); // Needs to recreate a spinbox of the KeyEdit.
```
Comment (Macksaur) [inline editor/animation_track_editor.cpp:7345]: Hello, can you explain/document this line better? It's causing the animation player to unselect keyframes when I click them in my fork. What is this line trying to solve? Thanks!

Comment (TokageItLab) [reply editor/animation_track_editor.cpp:7345]: The FPS in the key edit inspector is displayed dependent on the FPS of the snap value. For example, a key placed at 1s will appear as follows in FPS mode:

image image

I thought that changing the FPS with the keyedit inspector displayed would implicitly change the FPS in the inspector and have some negative effect, but sorry I did not realize that _update_snap_unit() is fired each key time is changed in the inspector (although I think it's a redundant).

There are some instability issues, such as undo after switching from second mode and keys not working due to spin box arrows, but this seems to be a problem that has been occurring for some time and probably needs to be revert in this area and stabilized from other parts.

**Pull request merged**